tsl/src/fdw/deparse.c

```diff
@@ -867,6 +867,60 @@ deparseDistinctClause(StringInfo buf, deparse_expr_cxt *context)
 			return;
 	}
 
+	if (query->hasDistinctOn)
+	{
+		/*
+		 * Pushing down DISTINCT ON is more complex than plain DISTINCT.
+		 * The DISTINCT ON columns must be a prefix of the ORDER BY columns.
+		 * Without this, the DISTINCT ON would return an unpredictable row
+		 * each time. There is a diagnostic for the case where the ORDER BY
+		 * clause doesn't match the DISTINCT ON clause, so in this case we
+		 * would get an error on the data node. There is no diagnostic for
+		 * the case where the ORDER BY is absent, so in this case we would
+		 * get a wrong result.
+		 * The remote ORDER BY clause is created from the pathkeys of the
+		 * corresponding relation. If the DISTINCT ON columns are not a prefix
+		 * of these pathkeys, we cannot push it down.
+		 */
+		int next_pathkey_index = 0;
+		foreach (l, query->distinctClause)
+		{
+			SortGroupClause *sgc = lfirst_node(SortGroupClause, l);
+			TargetEntry *tle = get_sortgroupclause_tle(sgc, query->targetList);
+
+			if (next_pathkey_index >= list_length(pathkeys))
+			{
+				return;
+			}
+
+			PathKey *pk = list_nth(pathkeys, next_pathkey_index);
+			EquivalenceClass *ec = pk->pk_eclass;
+			++next_pathkey_index;
+
+			/*
+			 * The find_ec_member_matching_expr() has many checks that don't seem
+			 * to be relevant here. Enumarete the pathkey EquivalenceMembers by
+			 * hand and find the one that matches the DISTINCT ON expression.
+			 */
+			EquivalenceMember *matching_em = NULL;
+			ListCell *ec_member_cell;
+			foreach (ec_member_cell, ec->ec_members)
+			{
+				EquivalenceMember *ec_member = lfirst_node(EquivalenceMember, ec_member_cell);
+				if (equal(ec_member->em_expr, tle->expr))
+				{
+					matching_em = ec_member;
+					break;
+				}
+			}
```

Comment:
I suggest moving this to an appropriately named function, similar to `find_ec_member_matching_expr()` but without the extra functionality that you refer to above.

Reply:
It's just one loop with one if. Even if I needed it somewhere else, I'd probably just write it anew. Too small/meaningless for a function.

Reply:
A function name also conveys what this code block does, so there's a readability aspect of trying to understand the code by "reading" it using aptly named functions as opposed to just doing it for code reuse purposes.

For example, that function you referred to called `find_ec_member_matching_expr()` made it immediately clear what happened here just by reading the name, while it is non-obvious from the code block what it does.

Reply:
There's a comment about this. It's more readable than a function name, and also I don't have to jump between screens if I want to see what is actually done. And I will want to see it, because the function names lie even more than the comments do. So extracting small functions is a net readability decrease, at least for my code reading habits.

> that function you referred to called find_ec_member_matching_expr() made it immediately clear what happened here just by reading the name

Yeah, and it looks like it's suitable for our use case, but in fact it's not, because it also does three unrelated checks useful for some other case. That's exactly what I'm talking about, there's no way to encode stuff like this into the function name, you have to read the comments/code anyway.